src/validation/parse.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 Helsing GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::Path;
+
+use crate::validation::data::{Packages, PackagesError};
+
+/// Errors parsing `buffrs` packages.
+#[derive(thiserror::Error, Debug)]
+#[allow(missing_docs)]
+pub enum ParseError {
+    #[error(transparent)]
+    Parse(#[from] anyhow::Error),
+    #[error(transparent)]
+    Adding(#[from] PackagesError),
+}
+
+/// Parser for `buffrs` packages.
+pub struct Parser {
+    parser: protobuf_parse::Parser,
+}
+
+impl Parser {
+    /// Create new parser with a given root path.
+    pub fn new(root: &Path) -> Self {
+        let mut parser = protobuf_parse::Parser::new();
+        parser.pure();
+        parser.include(root);
+        Self { parser }
+    }
+
+    /// Add file to be processed by this parser.
+    pub fn input(&mut self, file: &Path) {
+        self.parser.input(file);
+    }
+
+    /// Parse into [`Packages`].
+    pub fn parse(self) -> Result<Packages, ParseError> {
+        let fds = self.parser.file_descriptor_set()?;
+        let mut packages = Packages::default();
+        for file in &fds.file {
```

Comment:
This can be done using iterators

Reply:
I would `impl FromIterator` for `Packages` (@asmello did this in another PR and it is super elegant!):

https://doc.rust-lang.org/std/iter/trait.FromIterator.html

Reply:
I am aware of `FromIterator`, however we cannot implement `FromIterator` for `Packages` because the operation of adding is fallible. We need to be able to capture the error.

In my understanding, a more idiomatic way to write this would be using `try_fold()`, because fundamentally this operation is a folding operation. 

Something like:

```rust
fds.file.iter().try_fold(Packages::default(), |packages, item| package.add(item))?;
```

I might be wrong here, and not see something obvious tho.

Reply:
Okay i agree, fine for me to leave it like this.